rfcs/0000-libp2p.md

```diff
@@ -0,0 +1,149 @@
+# libp2p for coda
+
+## Summary
+[summary]: #summary
+
+Coda's networking today uses an [largely-unmaintained Kademlia library written in Haskell](https://github.com/codaprotocol/kademlia). We've been intending to replace it with some other maintained S/Kademlia implementation. We use Kademlia for peer discovery: finding IP address and ports of other protocol participants. On top of this, we build a "gossip net" for broadcasting messages to all participants with a simple flooding protocol. libp2p offers these same primitives plus some useful features like NAT traversal and connection relaying. It also has a Javascript implementation that works in the browser, which will be useful for our SDK.
+
+## Detailed Design
+[detailed-design]: #detailed-design
+
+libp2p itself is a large, flexible, extensible network stack with multiple implementations and multiple options for various functionality. I propose this minimal initial configuration:
+
+
+- On the daemon side, we will use `go-libp2p`. `rust-libp2p` lacks some features (pubsub validation) that we want.
+- For the transport, secio over TCP (optionally with WebSocket). secio is libp2p's custom transport security protocol. I believe the libp2p developers intend to replace secio with TLS 1.3 in the future. Transport security is not yet essential for us, but it's nice to have and will become more important once we start using relays.
+- libp2p multiplexes several protocols over a single transport connection. For the multiplexer, we will use the libp2p mplex protocol. We can easily add more in the future, but mplex is the only one implemented by `js-libp2p` at the moment.
+- Kademlia for peer discovery and routing.
+- "floodsub" for pubsub. More on this below.
+- A custom protocol for encapsulating Coda's Jane Street RPCs (possibly temporary).
+
+This basic configuration improves on our network stack in three major ways:
+
+1. Multiplexing over a single connection should slightly reduce our connection establishment overhead (we open a separate TCP stream per RPC right now).
+2. Transport security means connections are authenticated with the NodeID. Currently our only notion of node identity is an IP address.
+3. Browsers can join the DHT and connect to daemons.
+
+It still has some limitations: no NAT traversal, no browser↔browser connections, no message relaying. 
+ 
+Here's a WIP signature for the new net:
+
+```ocaml
```

Comment:
How does OCaml communicate with libp2p? Via FFI to Go?

Reply:
Using a simple (ad-hoc) JSON protocol over stdin/stdout of a subprocess. No FFI so we don't have to deal with any potential runtime conflicts.